### ISSUE TITLE
Missing ETag header when downloading files from s3proxy

### ISSUE THREAD

**Issue by bthdimension** [gaul/s3proxy#615]
When I do a streaming download using the aws adk, I get the following error:
`software.amazon.awssdk.core.exception.SdkClientException: Failed to send the request: Response missing required ETag header.`

The streaming is done as follows:
```
s3AsyncClient.getObject(
                r -> r.bucket(bucket).key(key),
                AsyncResponseTransformer.toBlockingInputStream());
InputStream inputStream = responseFuture.join();
IOUtils.copy(inputStream, fileOutputStream);
```

I'm running s3proxy within a JUnit test and set it up as follows:
```
        Properties properties = new Properties();
        properties.setProperty(S3ProxyConstants.PROPERTY_AUTHORIZATION, "aws-v2-or-v4");
        properties.setProperty(S3ProxyConstants.PROPERTY_IDENTITY, accessKeyId);
        properties.setProperty(S3ProxyConstants.PROPERTY_CREDENTIAL, secretAccessKey);
        properties.setProperty(S3ProxyConstants.PROPERTY_IGNORE_UNKNOWN_HEADERS, "true");
        properties.setProperty(S3ProxyConstants.PROPERTY_ENDPOINT, endpoint);
        properties.setProperty(FilesystemConstants.PROPERTY_BASEDIR, TestFileManagement.S3_BASE_DIRECTORY);

        context = ContextBuilder
                .newBuilder("filesystem")
                .credentials("identity", "credential")
                .overrides(properties)
                .build(BlobStoreContext.class);

        s3Proxy = S3Proxy.Builder.fromProperties(properties)
                .blobStore(context.getBlobStore())
                .build();
```
There are some randomly generated files in the S3_BASE_DIRECTORY.

My s3proxy version is 1.9.0 due to the issue mentioned here (see my last comment): https://github.com/gaul/s3proxy/issues/535
My aws sdk version is:

- software.amazon.awssdk:s3:2.24.3
- software.amazon.awssdk.crt:aws-crt:0.29.10

**Comment by gaul** [comment]
S3Proxy includes the ETag if it is exists.  For the filesystem backend, this is only present if S3Proxy created the object which stores the tag in the extended attributes if they are supported (I believe all platforms with modern Java versions).  If you are calling getBlob on an objected created outside S3Proxy, e.g., exposing an existing directory via S3, this attribute will not be present and thus S3Proxy will not include the header.  I am curious about `Response missing required ETag header.` -- this was not always a requirement so perhaps there is some regression in the AWS library or configuration you need to use?

**Comment by bthdimension** [comment]
Thank you very much for the quick answer. If I use the S3 client to upload the files, they indeed have the ETag. Is there a way to set these attributes manually? Otherwise I'll just use this workaround.
Thanks!

**Comment by gaul** [comment]
S3Proxy (via Apache jclouds) previously tried computing ETag on demand when it was not available in apache/jclouds@496e27f1afa32b90d0656c3f21387cf68a30cb31.  This was reverted in apache/jclouds@86e947ddedfa1a13976f9debeadae1da6f2f0190 since it was computationally expensive for large objects, particularly when listing objects.

Could you share why your library is enforcing the presence of ETag?  This should have a configuration knob to disable the behavior.

**Comment by bthdimension** [comment]
The libraries I'm using are 'software.amazon.awssdk:s3:2.24.3' and 'software.amazon.awssdk.crt:aws-crt:0.29.10'. The client that seems to be enforcing it is "software.amazon.awssdk.services.s3.S3AsyncClient" and it is specifically the CRT client.

I create the client as follows:
```
S3AsyncClient.crtBuilder()
                .endpointOverride(URI.create(url))
                .forcePathStyle(true)
                .credentialsProvider(StaticCredentialsProvider.create(credentials))
                .region(Region.US_EAST_1)
                .targetThroughputInGbps(targetThroughputInGbps)
                .minimumPartSizeInBytes(minimumPartSizeInBytes)
                .checksumValidationEnabled(false)
                .build()
```

When creating the CRT client, I' can't really find an option to disable the ETag check. I was only able to disable the checksum validation.

**Comment by zjl578254155** [comment]
> .forcePathStyle(true)

Hello, may I ask if this issue has been resolved?
I am creating the download file in this way, but there is still an error message saying that Etag is missing. I don't know where the problem lies?
<img width="832" alt="image" src="https://github.com/user-attachments/assets/cac5f5eb-8b56-456b-862b-33507bed38e9">

**Comment by gaul** [comment]
There is a way to fix the objects you already have, without re-uploading them from the client: copy each object onto itself through the S3 API.

```
aws --endpoint-url http://127.0.0.1:8080 s3api copy-object \
    --bucket bkt --key foreign.txt \
    --copy-source bkt/foreign.txt --metadata-directive REPLACE
```

or for a whole bucket, nested keys included:

```
aws --endpoint-url http://127.0.0.1:8080 s3 cp s3://bkt s3://bkt \
    --recursive --metadata-directive REPLACE
```

**Why it works.** On the filesystem backends `copyBlob` is a `getBlob` followed by a `putBlob` rather than a file copy, and `putBlob` hashes the payload as it writes and stores the digest in the `user.content-md5` extended attribute. So the copy creates the ETag the original file never had. It is safe to do in place: `putBlob` streams into a temporary file and only then moves it over the target, so the source is not truncated while it is being read.

Verified against master, with a file placed into the base directory rather than uploaded:

| | HEAD before | HEAD after |
| --- | --- | --- |
| `foreign.txt` | 200, no ETag header | 200, `ETag: "68a7d546a4bd40add9d1c90acb65c697"` |

Content unchanged, and the ETag matches the real MD5 of the bytes. The recursive form fixed `sub/nested.txt` too.

Three caveats:

* `--metadata-directive REPLACE` is required rather than decorative -- real S3 refuses a self-copy without it, so the flag keeps the recipe portable.
* It rewrites every object, so it costs a full read and write per object. That is the same work jclouds used to do implicitly and reverted as too expensive; here you pay it once, deliberately.
* It fixes the objects that exist when you run it. Files dropped into the directory afterwards still arrive without an ETag, so for a test fixture directory this belongs in the setup step -- or seed the fixtures through the S3 API in the first place, as @bthdimension ended up doing.

For the record on the original question: it is the **CRT** client specifically that requires the header (`S3AsyncClient.crtBuilder()`), because it validates that an object did not change between the ranged requests it issues in parallel. That is a different knob from `checksumValidationEnabled(false)`, which is why disabling checksums did not help. The plain async and sync clients do not insist.

@zjl578254155 your follow-up quotes `forcePathStyle`, which is unrelated to this -- the question that matters is whether those objects were written **through** S3Proxy or placed in its directory. If the latter, the copy above will fix them; if the former, please open a fresh issue, because that would be a different bug.

Closing as answered. Note that the behaviour itself is unchanged and deliberate: computing an ETag on demand was reverted upstream because it means reading every object, worst of all during a listing. Computing it lazily on GET and HEAD only, and caching it in the xattr, would be a reasonable middle path -- the backend already does exactly that shape of fallback for content type -- but that deserves its own issue rather than being buried here.
